### PR TITLE
Refactor IO following Cats

### PR DESCRIPTION
**Pull request by pakoito**
Following the changes in these diffs:
https://github.com/typelevel/cats-effect/pull/90
https://github.com/typelevel/cats-effect/pull/95

The second diff is still WIP so I'll keep an eye for any changes we can apply too.

### CODE REVIEW

File: kategory-effects/src/main/kotlin/kategory/effects/data/IO.kt [hunk 1]
```diff
@@ -211,138 +202,72 @@ sealed class IO<out A> : IOKind<A> {
...
-                    when (result) {
-                        is Right -> f(result.b).unsafeRunAsync(cb)
-                        is Left -> f.error(result.a, { RaiseError(it) }).unsafeRunAsync(cb)
-                    }
-                } catch (throwable: Throwable) {
-                    cb(Left(throwable))
-                }
-            }
+    override fun <B> map(f: (A) -> B): IO<B> =
+            // Allowed to do 32 map operations in sequence before
```
Comment (pakoito) [inline kategory-effects/src/main/kotlin/kategory/effects/data/IO.kt]: Quack

File: kategory-effects/src/main/kotlin/kategory/effects/data/internal/IOFrame.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package kategory.effects
+
+import kategory.Either
+
+interface IOFrame<in A, out R> : (A) -> R {
```
Comment (raulraja) [inline kategory-effects/src/main/kotlin/kategory/effects/data/internal/IOFrame.kt]: Any way to make these private or protected so they are not abused in the IO public API?

File: kategory-effects/src/main/kotlin/kategory/effects/data/internal/IORunLoop.kt [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+import kategory.left
+import kategory.right
+
+private typealias Current = IO<Any?>
+private typealias BindF = (Any?) -> IO<Any?>
+private typealias CallStack = ArrayStack<BindF>
+private typealias Callback = (Either<Throwable, Any?>) -> Unit
+
+@Suppress("UNCHECKED_CAST")
+object IORunLoop {
```
Comment (raulraja) [inline kategory-effects/src/main/kotlin/kategory/effects/data/internal/IORunLoop.kt]: Same as previous comment regarding visibility

Comment (pakoito) [reply kategory-effects/src/main/kotlin/kategory/effects/data/internal/IORunLoop.kt]: Fixed on Master

**Pull request merged**